File: src/integrations/documents/java/Archive.java
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Set;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

/** Package generated delivery names only; no client paths or shell execution. */
public final class Archive {
    private Archive() {}
    public static void main(String[] args) throws Exception {
        if (args.length < 3 || args.length % 2 != 1 || args.length > 1001) throw new IllegalArgumentException("Invalid entries");
        long total = 0;
        int pdfs = 0;
        Set<String> names = new HashSet<>();
        for (int i = 1; i < args.length; i += 2) {
            String name = args[i + 1];
            if (!name.matches("[A-Za-z0-9_-]+(?:/[A-Za-z0-9_-]+)*(?:\\.[A-Za-z0-9]+)?") || !names.add(name)) throw new IllegalArgumentException("Unsafe or duplicate name");
            total += Files.size(Path.of(args[i]));
            if (name.toLowerCase(java.util.Locale.ROOT).endsWith(".pdf")) pdfs++;
        }
        if (total > 100L * 1024 * 1024 || pdfs > 100) throw new IllegalArgumentException("Split this delivery");
        long copied = 0;
        try (ZipOutputStream zip = new ZipOutputStream(Files.newOutputStream(Path.of(args[0])))) {
            byte[] buffer = new byte[65536];
            for (int i = 1; i < args.length; i += 2) {
                ZipEntry entry = new ZipEntry(args[i + 1]);
                entry.setTime(0);
                zip.putNextEntry(entry);
                try (InputStream input = Files.newInputStream(Path.of(args[i]))) {
                    int count;
                    while ((count = input.read(buffer)) != -1) {
                        copied += count;
                        if (copied > 100L * 1024 * 1024) throw new IllegalArgumentException("Split this delivery");
                        zip.write(buffer, 0, count);
                    }
                }
                zip.closeEntry();
            }
        }
        System.out.println("{\"ok\":true}");
    }
}

File: src/integrations/documents/archive.ts
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { z } from 'zod';
import { runStage, type StageOptions } from './stage';

export type DeliveryArchiveEntry = { name: string; bytes: Uint8Array };
export const MAX_DELIVERY_BYTES = 100 * 1024 * 1024;

/** Only generated paths enter ZIP metadata. Validated again inside the JVM. */
export async function createDeliveryArchive(entries: DeliveryArchiveEntry[], options: StageOptions = {}): Promise<Buffer> {
  const names = new Set<string>();
  let total = 0;
  let pdfs = 0;
  if (entries.length === 0 || entries.length > 500) throw new Error('delivery_archive_limit');
  for (const entry of entries) {
    if (!/^[A-Za-z0-9_-]+(?:\/[A-Za-z0-9_-]+)*(?:\.[A-Za-z0-9]+)?$/.test(entry.name) || names.has(entry.name)) {
      throw new Error('delivery_archive_name');
    }
    names.add(entry.name);
    total += entry.bytes.byteLength;
    if (entry.name.toLowerCase().endsWith('.pdf')) pdfs++;
  }
  if (total > MAX_DELIVERY_BYTES || pdfs > 100) throw new Error('delivery_archive_limit');
  const work = await mkdtemp(join(tmpdir(), 'auditor-delivery-'));
  try {
    const output = join(work, 'delivery.zip');
    const args = [output];
    for (const [index, entry] of entries.entries()) {
      const path = join(work, `entry-${index}`);
      await writeFile(path, entry.bytes);
      args.push(path, entry.name);
    }
    const result = await runStage('Archive', args, z.object({ ok: z.literal(true) }), options);
    if (!result.ok) throw new Error(`delivery_archive_${result.failure.kind}`);
    return await readFile(output);
  } finally {
    await rm(work, { recursive: true, force: true });
  }
}
